artiq/coredevice/mirny.py

```diff
@@ -140,3 +149,160 @@ def write_ext(self, addr, length, data):
         if length < 32:
             data <<= 32 - length
         self.bus.write(data)
+
+
+class Almazny:
+    """
+    Almazny (High frequency mezzanine board for Mirny)
+
+    :param host_mirny - Mirny device Almazny is connected to
+    """
+
+    def __init__(self, dmgr, host_mirny):
+        self.mirny = dmgr.get(host_mirny)
+        self.mezz_data = 0
+
+    @kernel
+    def init(self):
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_REG_LATCH_BASE+0, 0),
+            (ALMAZNY_REG_LATCH_BASE+1, 0),
+            (ALMAZNY_REG_LATCH_BASE+2, 0),
+            (ALMAZNY_REG_LATCH_BASE+3, 0),
+            (ALMAZNY_SER_CLK, 0)
+            ])
+
+    @kernel
+    def att_to_mu(self, att):
+        """
+        Convert an attenuator setting in dB to machine units.
+
+        :param att: attenuator setting in dB [0-31.5] in 0.5 steps
+        :return: attenuator setting in machine units
+        """
+        if att > 31.5 or att < 0 or (att-int(att) != 0.0 and att-int(att) != 0.5):
+            raise ValueError("Invalid Almazny attenuator settings!")
+        return int(att * 2)
+
+    @kernel
+    def mu_to_att(self, att_mu):
+        """
+        Convert a digital attenuator setting to dB.
+
+        :param att_mu: attenuator setting in machine units
+        :return: attenuator setting in dB
+        """
+        return att_mu / 2
+
+    @kernel
+    def set_att_mu(self, channel, att_mu, output_on=True):
```

Comment:
There should be separate methods for the attenuators and RF switch, like the other cards.

Reply:
Unfortunately, we cannot set just the RF switch without setting the attenuators (and vice versa) - that's just how the shift register works. ...or I guess I could separate that if I keep the high-level internal state, and reapply one setting on change of the other.